### PR TITLE
First basic implementation of adding an image in a funkyheatmap

### PR DESCRIPTION
**Pull request by rcannood**

### CODE REVIEW

File: R/calculate_geom_positions.R [hunk 1]
```diff
@@ -154,22 +154,20 @@ calculate_geom_positions <- function(
...
   # gather image data
   img_data <- geom_data_processor("image", function(dat) {
     # If the location of the image is not provided using the "value"
     # column in the data, construct it from the directory and filename
-    if (!is.null(dat$extension)) {
+    print(dat$extension)
+    if (dat %has_name% "directory") {
       dat$value <- paste0(dat$directory, dat$value, ".", dat$extension)
-    }
+    }
```
Comment (rcannood) [inline R/calculate_geom_positions.R]: Print should probably be removed. Also, it might be better to process the directory and the extension separately, and to support having NA values in the column (for some reason).

```suggestion
    if (dat %has_name% "directory") {
      dat$value <- ifelse(is.na(dat$value) | is.na(dat$directory), dat$value, paste0(dat$directory, "/", dat$value))
    }
    if (dat %has_name% "extension") {
      dat$value <- ifelse(is.na(dat$value) | is.na(dat$extension), dat$value, paste0(dat$value, ".", dat$extension))
    }
```

File: R/calculate_geom_positions.R [hunk 2]
```diff
@@ -155,6 +154,24 @@ calculate_geom_positions <- function(
...
   })
 
+  # gather image data
+  img_data <- geom_data_processor("image", function(dat) {
+    # If the location of the image is not provided using the "value"
+    # column in the data, construct it from the directory and filename
+    print(dat$extension)
+    if (dat %has_name% "directory") {
+      dat$value <- paste0(dat$directory, dat$value, ".", dat$extension)
+    }
```
Comment (rcannood) [inline R/calculate_geom_positions.R:164]: Print should probably be removed. Also, it might be better to process the directory and the extension separately, and to support having NA values in the column (for some reason).

```suggestion
    if (dat %has_name% "directory") {
      dat$value <- ifelse(is.na(dat$value) | is.na(dat$directory), dat$value, paste0(dat$directory, "/", dat$value))
    }
    if (dat %has_name% "extension") {
      dat$value <- ifelse(is.na(dat$value) | is.na(dat$extension), dat$value, paste0(dat$value, ".", dat$extension))
    }
```

File: R/compose_ggplot.R [hunk 3]
```diff
@@ -168,6 +168,17 @@ compose_ggplot <- function(
       size = .25
     )
   }
+  # PLOT IMAGES
+  if (nrow(geom_positions$img_data) > 0) {
+    for(r in 1:nrow(geom_positions$img_data)) {
```
Comment (rcannood) [inline R/compose_ggplot.R:173]: Use `seq_len()` instead of `1:nrow(...)`

```suggestion
    for (r in seq_len(nrow(geom_positions$img_data))) {

File: R/funky_heatmap.R [hunk 4]
```diff
@@ -50,6 +50,10 @@
 #'     (only for `geom = "text"`).
 #'
 #' * `label`: Which column to use as a label (only for `geom = "text"`).
+#' 
+#' * `directory`: Which directory to use to find the images (only for `geom = "image"`).
+#' 
+#' * `extension`: The extension of the images (only for `geom = "image"`).
```
Comment (rcannood) [inline R/funky_heatmap.R:56]: Please update the man/* files after changing the documentation ;)

File: tests/testthat/test-image_geom.R [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+g_path <- funky_heatmap(
+  data = data_path,
+  column_info = column_info_data_path,
+  column_groups = column_groups,
+  row_info = row_info,
+  row_groups = row_groups,
+  palettes = palettes,
+  expand = list(xmax = 4)
+)
+ggsave("test_data_path.pdf", g, device = "pdf")
```
Comment (rcannood) [inline tests/testthat/test-image_geom.R:84]: No need to write the pdfs to file, IMO. Also, this file does not contain any actual tests.

There's also a lot of additional code which is not really relevant to the test (though perhaps edge cases can occur with all of the extra data, who knows).

I'd suggest:

```suggestion
test_that("image geom with full paths", {
  data <- mtcars %>%
    rownames_to_column("id") %>%
    arrange(desc(mpg)) %>%
    head(20) %>%
    mutate(type = c(rep("test_images/ice.png", 10), rep("test_images/electric.png", 10)))

  column_info <- tribble(
    ~id,     ~group,         ~name,                      ~geom,        ~palette,    ~options,
    "id",    NA,             "",                         "text",       NA,          list(hjust = 0, width = 6),
    "mpg",   NA,             "Miles / gallon",           "bar",        NA,          list(width = 4, legend = FALSE),
    "type",  NA,             "Type of engine",           "image",      NA,          lst(directory = "test_images/", extension = "png")
  )
  
  g <- funky_heatmap(
    data = data,
    column_info = column_info,
    expand = list(xmax = 4)
  )
  expect_true(ggplot2::is.ggplot(p))
})

test_that("image geom with directory and extension options", {
  data <- mtcars %>%
    rownames_to_column("id") %>%
    arrange(desc(mpg)) %>%
    head(20) %>%
    mutate(type = c(rep("ice", 10), rep("electric", 10)))

  column_info <- tribble(
    ~id,     ~group,         ~name,                      ~geom,        ~palette,    ~options,
    "id",    NA,             "",                         "text",       NA,          list(hjust = 0, width = 6),
    "mpg",   NA,             "Miles / gallon",           "bar",        NA,          list(width = 4, legend = FALSE),
    "type",  NA,             "Type of engine",           "image",      NA,          lst(directory = "test_images/", extension = "png")
  )
  
  g <- funky_heatmap(
    data = data,
    column_info = column_info,
    expand = list(xmax = 4)
  )
  expect_true(ggplot2::is.ggplot(p))
})
```

Ideally, you could add more checks to ensure that indeed an image is being plotted with cowplot.

File: vignettes/mtcars_images.Rmd
Comment (rcannood) [inline vignettes/mtcars_images.Rmd:1]: Please include this in the standard mtcars vignette instead of creating a separate vignette for just one geom ;)

File: tests/testthat/test-image_geom.R (hunk 5 above)
Comment (LouiseDck) [reply tests/testthat/test-image_geom.R]: I adapted the tests to this format, thanks! However, the use of `NA` caused this assert `is.character(column_info$palette) | is.factor(column_info$palette)` in `verify_column_info.R` to fail. I circumvented the issue by using `NA_character_`,  but maybe either this assert, or the if statement before this assert should be changed?

Comment (rcannood) [reply tests/testthat/test-image_geom.R]: Good point!

While I think `NA_character_` is the 'better' option here, I agree `NA` should also work.

I think it'd be best to add a statement before the assert that converts any NA to NA_character_. Or just add `is.na(column_info$palette) |` to the assert.

**Review by rcannood — COMMENTED** [review]
LGTM!
